lint/capability_fields_analyser.go

```diff
@@ -0,0 +1,140 @@
+/*
+ * Cadence-lint - The Cadence linter
+ *
+ * Copyright 2019-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lint
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/tools/analysis"
+)
+
+func DetectCapabilityType(typeToCheck ast.Type, structWithPubCapabilities map[string]struct{}) bool {
```

Comment:
Agreed, will do.

Not cause I am nitpicking, but just to clarify -

The idea is that contract is checked as a whole and once it has a "leak" warning is issued.
However, for structs and resources it's different. It's fine to have a struct with a public capability field if this struct is always used to declare private fields in a contract.

For this reason a map of structWithPubCapabilities is constructed. It will allow to see if those structs are used to declare public fields. For contracts - we won't include them in this map as once there is a contract with a leak we already issue a warning. 

TBH I dont know if contract can be embedded in another contract. Seems, like this is not the case in Cadence. So, also a reason why contract shouldn't be in this map. Pls correct my understanding , my kwnoledge of Cadence is on a beginner level.